libraries/fc/src/network/http/http_client.cpp

```diff
@@ -0,0 +1,361 @@
+#include <fc/network/http/http_client.hpp>
+#include <fc/io/json.hpp>
+#include <fc/scoped_exit.hpp>
+
+#include <boost/beast/core.hpp>
+#include <boost/beast/http.hpp>
+#include <boost/beast/version.hpp>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/ssl/error.hpp>
+#include <boost/asio/ssl/stream.hpp>
+
+using tcp = boost::asio::ip::tcp;       // from <boost/asio/ip/tcp.hpp>
+namespace http = boost::beast::http;    // from <boost/beast/http.hpp>
+namespace ssl = boost::asio::ssl;       // from <boost/asio/ssl.hpp>
+
+namespace fc {
+
+class http_client_impl {
+public:
+   using host_key = std::tuple<std::string, std::string, uint16_t>;
+   using raw_socket_ptr = std::unique_ptr<tcp::socket>;
+   using ssl_socket_ptr = std::unique_ptr<ssl::stream<tcp::socket>>;
+   using connection = static_variant<raw_socket_ptr, ssl_socket_ptr>;
+   using connection_map = std::map<host_key, connection>;
+   using error_code = boost::system::error_code;
+   using deadline_type = boost::posix_time::ptime;
+
+   http_client_impl()
+   :_ioc()
+   ,_sslc(ssl::context::sslv23_client)
+   {
+      set_verify_peers(true);
+   }
+
+   void add_cert(const std::string& cert_pem_string) {
+      error_code ec;
+      _sslc.add_certificate_authority(boost::asio::buffer(cert_pem_string.data(), cert_pem_string.size()), ec);
+      FC_ASSERT(!ec, "Failed to add cert: ${msg}", ("msg", ec.message()));
+   }
+
+   void set_verify_peers(bool enabled) {
+      if (enabled) {
+         _sslc.set_verify_mode(ssl::verify_peer);
+      } else {
+         _sslc.set_verify_mode(ssl::verify_none);
+      }
+   }
+
+   template<typename SyncReadStream, typename Fn, typename CancelFn>
+   error_code sync_do_with_deadline( SyncReadStream& s, deadline_type deadline, Fn f, CancelFn cf ) {
+      optional<error_code> timer_result;
+      boost::asio::deadline_timer timer(s.get_io_service());
+
+      timer.expires_at(deadline);
+      bool timer_cancelled = false;
+      timer.async_wait([&timer_result, &timer_cancelled] (const error_code& error) {
+         if (!error && !timer_cancelled) {
```

Comment:
Had to read that a couple of times and apply De Morgan's Law to understand. Would be easier to read as `if (error || timer_cancelled)`. Ah but maybe you were avoiding optional bool operator.  `if (!!error || timer_cancelled)` might still be better but then it is much more subjective.

Reply:
So, it looks odd for sure.   I think I will change the way this works because there are only 2 options for the invocation of this according to boost:

call with a "success error" meaning the timer successfully expired OR `boost::asio::error::operation_aborted` meaning it was canceled when not in-flight.  the `timer_cancelled` bool handles canclations when the success is already in flight.  

All the code cares about is whether the timer successfully expired.  So, only when this is called with a "success error" (aka !error) and we havent set the cancelled flag, should we set the timer_result to something ...